File: src/main/scala/com/github/mmolimar/kukulcan/KConsumer.scala
package com.github.mmolimar.kukulcan

import java.util.Properties

import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.tools.{ToolsUtils => JToolsUtils}

import scala.collection.JavaConverters._

private[kukulcan] object KConsumer extends Api[KConsumer[AnyRef, AnyRef]]("consumer") {

  override protected def createInstance(props: Properties): KConsumer[AnyRef, AnyRef] = {
    KConsumer[AnyRef, AnyRef](props)
  }

}

private[kukulcan] case class KConsumer[K, V](props: Properties) extends KafkaConsumer[K, V](props) {

  import org.apache.kafka.common.{Metric, MetricName}

  def reload(): Unit = KConsumer.reload()

  def getMetrics(groupRegex: String = ".*", nameRegex: String = ".*"): Map[MetricName, Metric] = {
    metrics.asScala
      .filter(metric => metric._1.group.matches(groupRegex) && metric._1.name.matches(nameRegex))
      .toMap
  }

  def listMetrics(groupRegex: String = ".*", nameRegex: String = ".*"): Unit = {
    JToolsUtils.printMetrics(getMetrics(groupRegex, nameRegex).asJava)
  }

}

File: src/main/java/com/github/mmolimar/kukulcan/java/KConsumer.java
package com.github.mmolimar.kukulcan.java;

import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.Metric;
import org.apache.kafka.common.MetricName;
import org.apache.kafka.tools.ToolsUtils;

import java.util.Collections;
import java.util.Map;
import java.util.Properties;
import java.util.stream.Collectors;

public class KConsumer<K, V> extends KafkaConsumer<K, V> {

    public KConsumer(Properties props) {
        super(props);
    }

    public void subscribe(String topic) {
        subscribe(Collections.singletonList(topic));
    }

    public Map<MetricName, Metric> getMetrics() {
        return getMetrics(".*");
    }

    public Map<MetricName, Metric> getMetrics(String groupRegex) {
        return getMetrics(groupRegex, ".*");
    }

    public Map<MetricName, Metric> getMetrics(String groupRegex, String nameRegex) {
        return metrics().entrySet().stream()
                .filter(metric -> metric.getKey().group().matches(groupRegex) &&
                        metric.getKey().name().matches(nameRegex))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    public void listMetrics() {
        listMetrics(".*");
    }

    public void listMetrics(String groupRegex) {
        listMetrics(groupRegex, ".*");
    }

    public void listMetrics(String groupRegex, String nameRegex) {
        ToolsUtils.printMetrics(getMetrics(groupRegex, nameRegex));
    }
}
